### ISSUE TITLE
URI.withinString does not parse URIs that contain balanced parentheses

### ISSUE THREAD

**Issue by astorije** [medialize/URI.js#247]
These 2 examples are correctly parsed by GitHub so this is a good place to demonstrate:
1. https://example.com/with_(balanced_parentheses)
2. https://example.com/with_unbalanced_parenthesis)

`URI.withinString` will get case **1** correctly, but not case **2**.

Consider this real-life example: http://en.wikipedia.org/wiki/The_Martian_(Andy_Weir), here is what happens using the latest version on npm (`URIjs@1.16.0`):

``` js
> URI.withinString('https://en.wikipedia.org/wiki/The_Martian_(Weir_novel)', function (url) { return '<a>' + url + '</a>'; });
'<a>https://en.wikipedia.org/wiki/The_Martian_(Weir_novel</a>)'
```

The second parenthesis gets pushed out.

Other bracket types (`[] {} <>') should follow the same rule.

**Comment by rodneyrehm** [comment]
Nice catch! The problem looks simple enough to allow someone else to try fixing it. Here are some pointers to help with that:

[URI.withinString](https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L884) uses the [URI.findUri](https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L213-L220) expressions to hone in on potential URLs. The expression [URI.findUri.trim](https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L219) is used to [trim those characters](https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L908), because [URI.findUri.end](https://github.com/medialize/URI.js/blob/gh-pages/src/URI.js#L217) is only looking for some whitespace.

I guess I would try splitting `URI.findUri.trim` into opening and closing characters and only trim closing chararcters, if the identifed URI does not contain the corresponding opening character. Adding a [test](https://github.com/medialize/URI.js/blob/gh-pages/test/test.js#L1506-L1556) for the change would likely help, too.

**Comment by astorije** [comment]
Hi @rodneyrehm, thanks for the pointers! I'll see if I can send a PR within the next few days and/or come back with questions. But yes, my tests were already written when your answer came in :-)

**Comment by blaenk** [comment]
Any news on this?

**Comment by davericher** [comment]
bump

**Comment by rodneyrehm** [comment]
since there was no patch provided by the community, I patched it myself… 

The characters `()`, `[]`, `{}` and `<>` are considered and the closing character is included in the URL if it also contains the opening character:

```
http://example.com/foo(bar)
-> "http://example.com/foo(bar)"

http://example.com/foo(bar))
-> "http://example.com/foo(bar)"

http://example.com/foobar))
-> "http://example.com/foobar"
```

released as [v1.18.4](https://github.com/medialize/URI.js/releases/tag/v1.18.4).

**Comment by astorije** [comment]
Thanks a lot, @rodneyrehm! Sorry I wasn't helpful on that.
